### ISSUE TITLE
Firing notifies from within an inline'd LWRP provider.

### ISSUE THREAD

**Issue by docwhat** [chef/chef#3575]
Currently, it is impossible to do this:

```
# recipes/default.rb

service 'foo'
```

```
# provider/myprovider.rb
use_inline_resources

action :something do
  file '/tmp/foo' do
    notifies :restart, 'service[foo]'
  end
end
```

It returns an error because it cannot find the resource `service[foo]` because the LWRP provider runs in a temporary `run_context`, not the `saved_run_context` or the "parent" run_context.

I feel like this should be able to notify/subscribe "up" (e.g. to parent `run_context`s), but not down (e.g. into the resources *hidden* by an LWRP).

If I'm going about this wrong, please tell me.  Initially, this seemed like the "correct" thing to do; then I dug around the code and realized what was going on.

Suggestion:

Perhaps change [lwrp_base.rb`'s line 59](https://github.com/chef/chef/blob/master/lib/chef/provider/lwrp_base.rb#L59) like so:

```
diff --git a/lib/chef/provider/lwrp_base.rb b/lib/chef/provider/lwrp_base.rb
index b5efbb2..d434a49 100644
--- a/lib/chef/provider/lwrp_base.rb
+++ b/lib/chef/provider/lwrp_base.rb
@@ -56,7 +56,7 @@ def recipe_eval_with_update_check(&block)
           saved_run_context = @run_context
           temp_run_context = @run_context.dup
           @run_context = temp_run_context
-          @run_context.resource_collection = Chef::ResourceCollection.new
+          @run_context.resource_collection = saved_run_context.resource_collection

           return_value = instance_eval(&block)
           Chef::Runner.new(@run_context).converge
```

Though thinking about, I think we want a *copy* of the `resource_collection`, not the actual `resource_collection` from the parent. Otherwise you'd get a loop or something. :-/

**Comment by docwhat** [comment]
Yeah, tried it and it loops forever. :-/

**Comment by ranjib** [comment]
dont use inline resources if you dont want that behavior. inline resources are isolated, if you want to notify   resources outside the provider body, do so from the consumer of the lwrp. if you want that to happen inside the provider body bring the service resource declaration inside the provider action.  current implementation allows automatic management of resource's updated_by_last_action  stuff. 
we can extend the lwrp api to setup some automatic notification as well, if this is a common use case.

**Comment by docwhat** [comment]
I suspect it is common enough use case.  It certainly *looks* like it should work.  I'd argue that since it does work elsewhere it should work here or alternatively raise a really helpful message (e.g. "you're using inline resources and notifies/subscribes can't talk out of the provider").

> if you want that to happen inside the provider body bring the service resource declaration inside the provider action.

I was under the impression that this is bad, because you end up creating multiple versions of the resource... using my example from above, you end up with `service 'foo'` being created once per invocation of the resource/provider.  Since these are in the inline resource, they will collide... right?  I don't get any warnings about cloning resources like I normally do.

Trying this in a Heavy Weight provider, I'm not getting the notifies to trigger at all...

```
      def action_create
        file_text = diamond_collector_config
        config_file = Resource::File.new(file_path, run_context)
        config_file.mode 0644
        config_file.content file_text
        config_file.action :nothing

        # Try 1
        config_file.notifies :restart, 'runit_service[diamond]'

        # Try 2
        runit_service_diamond = run_context.resource_collection.find(runit_service: 'diamond')
        runit_service_diamond.subscribes :restart, config_file

        config_file.run_action(:create)

        new_resource.updated_by_last_action(config_file.updated_by_last_action?)
      end
```

Neither one is triggering a restart for `runit_service[diamond]`.

**Comment by docwhat** [comment]
Huh... okay, to make it work in the heavy-weight provider, I had to do this:

```
      def action_create
        # It didn't work with the string. I got the error:
        # NoMethodError: undefined method `run_action' for "runit_service[diamond]":String
        new_resource.notifies :restart, run_context.resource_collection.find(runit_service: 'diamond')

        file_text = diamond_collector_config
        config_file = Resource::File.new(file_path, run_context)
        config_file.mode 0644
        config_file.content file_text
        config_file.action :nothing
        config_file.run_action(:create)

        new_resource.updated_by_last_action(config_file.updated_by_last_action?)
      end
```

Something is weird here.  `notifies` isn't working the same in different places.  Obviously my try 1 and 2 above weren't sending things correctly (or sending them into the queue too late?).

**Comment by lamont-granquist** [comment]
So the stock answer to this is either you need to move more resources into your LWRP or make it the responsibility of the user to setup the notifies.  There's also the common case where what what you really want to do is use a definition to setup an accumulator on a template resource which is in the parent run context (because its a definition there is no child run context) and then use that to signal some service restart or whatever when the template updates (e.g. 
https://github.com/lamont-cookbooks/sk_sysctl/blob/master/definitions/sysctl.rb)

If you've got a more compelling use case you need to spell it out completely, otherwise this is a FAQ and needs to be closed.

**Comment by ranjib** [comment]
im closing this. its as design. we are trying to combine the recipe DSL features within the LWRP/HWRP, this is possible but you have to use low level chef api, and its likely to be error prone. Both me and @lamont-granquist given some ideas that can be worked on (like moving the service resource declaration inside lwrp , or not using inline resources). @docwhat we can continue the discussion, and come up with some workaround or a better solution, but theres nothing actionable on chef client end, hence im closing this.

**Comment by docwhat** [comment]
I ended up doing this in my heavy provider:

``` ruby
      def action_create
        new_resource.notifies :restart, run_context.resource_collection.find(runit_service: 'diamond')

        file_text = diamond_collector_config
        config_file = Resource::File.new(file_path, run_context)
        config_file.mode 0644
        config_file.content file_text
        config_file.action :nothing
        config_file.run_action(:create)

        new_resource.updated_by_last_action(config_file.updated_by_last_action?)
     end
```

Adding the notifies to `new_resource` seems to work since I ensure that `new_resource.updated_by_last_action()` is set correctly.

Is this acceptable? or is this prone to break?  I wouldn't be able to do this in a LW provider because I wouldn't have access to the correct `run_context`.

**Comment by ranjib** [comment]
i dont think its prone to break, as you are mostly using the public api. but i see you are not using converge_by blocks,, so it might not be whyrun safe.

**Comment by lamont-granquist** [comment]
It won't break, but its fairly bad practice and tight coupling.  Now your HWRP can't be used with any other kind of service resource other than the runit_service, while you still have the external requirement to add the `runit_service[diamond]` to the run list.

**Comment by docwhat** [comment]
@lamont-granquist True.  I don't care about that (yet).  I'm trying to get this work in our system.  Then I'll break it out and open source it.  I'll try breaking out the service at that time.

@ranjib I didn't think I needed converge_by blocks if I used another resource.  Is that not true?  This way I'll see a message about creating/not-creating the config file.  Otherwise I just see a message about the `diamond_collector` resource.

**Comment by docwhat** [comment]
BTW: Both of you, thanks for the help.

**Comment by coderanger** [comment]
Worth noting that the Poise `notifying_block` does not have this issue.

**Comment by docwhat** [comment]
@coderanger You mean this? https://github.com/poise/poise#notifying-block

**Comment by lamont-granquist** [comment]
I'm still unclear of why a well-written set of cookbooks would need to use this.  Without the compelling use case being explained it seems like a way to enable writing bad cookbooks tightly coupled to the internals of other resources.

**Comment by docwhat** [comment]
@lamont-granquist 

I'm open to the argument that I'm writing a bad cookbook, but I'm not seeing a replacement pattern or strategy to use instead.

The use case is:
  * I have a bit of code that sets things up with a configuration (e.g. a complicated `runit_service` block).  This code makes sense (to me) to just put in a default recipe.
  * I have incremental configuration additions (e.g. a bunch of files dropped into a directory that the service will use for configuration.)  These additions make sense to be an RP of some kind due to translation (in the above case, ruby -> python).  But each change should do a delayed notification to the first bit.

Every time I tried to make the service part of the RP, I'd get all these warnings about cloning resources, etc. I've tried using the `resources()` block to make sure I'm using the same resource but that never works.

So how would you do this?  What's a "good" way to do this then?

**Comment by lamont-granquist** [comment]
what is wrong with hanging the delayed notification off of the RP call rather than internal to RP?  if you just don't like typing the notification line all the time, wrap it with a definition to avoid having to type it all the time.  that keeps the runit_service in the main run_context along with the RPs and all the notifications.  then internal to the RPs you don't worry about notifications.

and if there's problems caused by using definitions to do compile-time "macro preprocessing" then we should fix that use case, or maybe create some way to inject a default notification into an RP that you write just like default resource property values.

**Comment by lamont-granquist** [comment]
I did actually take the time to read notifying_block now and I think I understand it, and we could simply write an RFC to make use_inline_resources type providers behave that way (optionally, or possibly by default).   I can squint my eyes and sort of see that what you're doing is similar to variable scope -- as long as there's no change the syntax and notifies bubble up to the main run_context like that.

**Comment by coderanger** [comment]
The use case is generally notifying a service that something related to it has changed (config file, service def, etc). This doesn't belong on the LWRP invocation because it needs to be present 100% of the time for correct operation and even suggesting that level of silently-failing boilerplate code is a non-starter for my cookbooks. For example, `httpd_service` and `httpd_config`. Definitions _need_ to die, so they are not a solution, and even then that would mean having `httpd_config` be both a resource and definition and that sounds even more confusing.

Parent-based lookups between resource collections has had no ill effects in Poise, so I'm unclear why people are so resistant to it. It follows the normal Ruby (and every other language) rules of you can see out of a closure but not back in. If you want to not have external lookups, that still an option, just make sure your code does that.

**Comment by docwhat** [comment]
FYI: I ran into this again today... I wrote a provider that bundles `docker_image` and a custom replacement for `docker_container` together: `custom_docker_service`

If I try to run multiples (for scaling and redundancy) then the second and higher `docker_container`s won't reload when the `docker_image` changes.

It's  quite possible I'm bundling the wrong stuff. I wrote this a while ago.  I could add a way to scale this automagically, but that seems overly difficult and makes container naming pre-determined.

Pseudo code:

```
action(:create) do
  docker_image resource.full_image_name do
     ..
  end

  service resource.name do
    ...
    subscribes :restart, "docker_image[#{resource.full_image_name}]"
  end
end
```

This is because each `docker_image` is actually separate, which is unexpected.

I hope this use-case is helpful.
